### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by kikooOOoo16**
Initial version of the chat API.
Setup .env-cmdrc in root with the following code 

  "development": {
    "NG_APP_URL": "http://localhost:4200",
    "MONGODB_URL": "mongodb://localhost:27017/nodeChatDevDB",
    "JWT_SECRET": "secret string"
  },

npm run dev

Or download both the Node and Angular apps inside same parent directory and use docker-compose build and docker-compose up.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,6 +15,45 @@ pids
 *.seed
 *.pid.lock
 
+# Idea files
+.idea/workspace
+.idea/*.iws
+.idea/workspace.xml
+*/workspace.xml
+
+# User-specific stuff
```
Comment (slobodan-tonchevski) [inline .gitignore]: You can ignore everything inside .idea/* at once

File: .gitignore [hunk 2]
```diff
@@ -15,6 +15,45 @@ pids
...
+.idea/**/workspace.xml
+.idea/**/tasks.xml
+.idea/**/usage.statistics.xml
```
Comment (slobodan-tonchevski) [inline .gitignore]: Idea files shouldnt be pushed at all

File: app/custom.d.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import {User} from "./src/interfaces/user";
```
Comment (slobodan-tonchevski) [inline app/custom.d.ts:1]: Why the name of the file is custom ?

File: app/server.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    if (isNaN(port)) {
+        // named pipe
+        return val;
+    }
+
+    if (port >= 0) {
+        //  port num
+        return port;
+    }
+    return false;
```
Comment (slobodan-tonchevski) [inline app/server.ts]: Why do we return false ? What happens if you return false sa port ?

File: app/server.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+const onError = (error: { syscall: string; code: any; }) => {
+    if (error.syscall !== 'listen') {
+        throw error;
+    }
+    let addr;
```
Comment (slobodan-tonchevski) [inline app/server.ts]: Why the addr is let ?  Is it reassigned somewhere

File: app/src/app.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+// parse incoming JSON
+app.use(express.json());
+
+// parse cookie data between server and client
+app.use(cookieParser())
+
+// Allow CORS communication
+app.use((req, res, next) => {
+    res.setHeader('Access-Control-Allow-Origin', 'http://localhost:4200');
```
Comment (slobodan-tonchevski) [inline app/src/app.ts]: http://localhost:4200' Shouldnt be hardcoded it should be coming from env file

File: app/src/chat/exceptions/custom-exception-type.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+export enum customExceptionType {
+    unauthorizedAction = 'UNAUTHORIZED_ACTION',
```
Comment (slobodan-tonchevski) [inline app/src/chat/exceptions/custom-exception-type.ts]: UNAUTHORIZED_ACTION = 'UNAUTHORIZED_ACTION',
UNAUTHORIZED_ACTION_NOT_ROOM_AUTHOR = 'UNAUTHORIZED_ACTION_NOT_ROOM_AUTHOR',
EXPIRED_USER_TOKEN = 'EXPIRED_USER_TOKEN'

etc

It should be like: And allso the enum name should be starting with Upper case


export enum UserFlagsEnum {
  TERMS_AND_CONDITIONS_CONSENT = 'TERMS_AND_CONDITIONS_CONSENT',
  PRIVACY_POLICY_CONSENT = 'PRIVACY_POLICY_CONSENT',
}

File: app/src/chat/exceptions/custom-exception.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+export abstract class CustomException extends Error {
```
Comment (slobodan-tonchevski) [inline app/src/chat/exceptions/custom-exception.ts]: I guess a better name is AbstractException or Base Exception

File: app/src/chat/exceptions/custom-exception.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+
+    protected constructor(message: string) {
+        super(message);
+        Object.setPrototypeOf(this, CustomException.prototype);
```
Comment (slobodan-tonchevski) [inline app/src/chat/exceptions/custom-exception.ts]: Why do we need to set prototype ? What is the purpose of it

File: app/src/chat/exceptions/exception-factory.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+import {ProblemUpdatingRoom} from "./problem-updating-room";
+import {ProblemDeletingRoom} from "./problem-deleting-room";
+import {ProfaneLanguageNotAllowed} from "./profane-language-not-allowed";
+import {UnauthorizedActionNotRoomAuthor} from "./unauthorized-action-not-room-author";
+import {ProblemSavingUserSocketId} from "./problem-saving-user-socket-id";
+
+export class ExceptionFactory {
+    static createException = (exceptionType: customExceptionType)  => {
+        switch (exceptionType) {
+            case customExceptionType.expiredUserToken:
```
Comment (slobodan-tonchevski) [inline app/src/chat/exceptions/exception-factory.ts]: Instead using switch you can create a map to do this and its more readable

You can do something like:

const exceptionMap = {
    BaseException.EXPIRED_USER_TOKEN => ExpiredUserToken,
    ....
};

return new exceptionMap[BaseException.EXPIRED_USER_TOKEN];

File: app/src/chat/exceptions/expired-user-token.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+    constructor() {
+        super(`User token has expired.`);
+        Object.setPrototypeOf(this, ExpiredUserToken.prototype);
+    }
+
+    printError = (): string => {
+        return 'Error: User token has expired.';
+    }
+
```
Comment (slobodan-tonchevski) [inline app/src/chat/exceptions/expired-user-token.ts]: You don't need an empty line here I assume

File: app/src/chat/exceptions/expired-user-token.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import {CustomException} from "./custom-exception";
+
+export class ExpiredUserToken extends CustomException {
+
+    constructor() {
+        super(`User token has expired.`);
+        Object.setPrototypeOf(this, ExpiredUserToken.prototype);
```
Comment (slobodan-tonchevski) [inline app/src/chat/exceptions/expired-user-token.ts]: I don't understand why do we need printError message and super message

File: app/src/chat/exceptions/invalid-message-sent.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import {CustomException} from "./custom-exception";
+
+export class InvalidMessageSent extends CustomException {
+
+    constructor() {
```
Comment (slobodan-tonchevski) [inline app/src/chat/exceptions/invalid-message-sent.ts]: Same for all of these

File: app/src/chat/message-generator-service.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
+import {Message} from "../interfaces/message";
+import {User} from "../interfaces/user";
+
+export class MessageGeneratorService {
+    private static instance: MessageGeneratorService;
+
+    private constructor() {
```
Comment (slobodan-tonchevski) [inline app/src/chat/message-generator-service.ts]: If there is not anything in the constructor I guess it can be deleted

File: app/src/chat/message-generator-service.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    public static getInstance(): MessageGeneratorService {
```
Comment (slobodan-tonchevski) [inline app/src/chat/message-generator-service.ts]: You could add return type and input param comments on each method

File: app/src/chat/message-generator-service.ts [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        if (!MessageGeneratorService.instance) {
+            MessageGeneratorService.instance = new MessageGeneratorService();
+        }
+        return MessageGeneratorService.instance;
+    }
+
+    public generateMessage = (user: User | undefined, msgText: string): Message => {
+        let author;
```
Comment (slobodan-tonchevski) [inline app/src/chat/message-generator-service.ts]: const author = {
    id: user._id ??  '',
    name: user.name ?? 'Server'
}

File: app/src/chat/rooms-service.ts [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+import Filter from "bad-words";
+
+export class RoomsService {
+    private static instance: RoomsService;
+    private customException!: CustomException;
+
+    private constructor() {
+    }
+
+    public static getInstance(): RoomsService {
```
Comment (slobodan-tonchevski) [inline app/src/chat/rooms-service.ts]: I guess that we need factory for creating these instances. So you can add it in separate rooms service factory and generate this service from there

File: app/src/chat/rooms-service.ts [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+        if (!RoomsService.instance) {
+            RoomsService.instance = new RoomsService();
+        }
+        return RoomsService.instance;
+    }
+
+    // Create new room
+    createRoom = async (currentUser: User, newRoom: Room): Promise<{ roomName: string | undefined, createRoomErr: string }> => {
```
Comment (slobodan-tonchevski) [inline app/src/chat/rooms-service.ts]: Why did you add functions like this  ?

What about sth like this for example

  async autoLoginURL(@JwtToken() jwtToken: string): Promise<string> {
    let userUUID = null;
    if (jwtToken) {
      const payload = await this.authService.getTokenPayload(jwtToken);
      userUUID = payload.userUUID;
    }
    return this.authService.getAutoLoginURL(userUUID);
  }

Comment (slobodan-tonchevski) [inline app/src/chat/rooms-service.ts]: The room you are returning can be added to a separate object RoomDTO 

And then return Promise<RoomDTO>

File: app/src/chat/rooms-service.ts [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+        let createRoomErr = '';
+        // check if all data provided and is valid for newRoom
+        if (newRoom.name === '' || newRoom.description === '' || newRoom.description.length < 10) {
+            Logger.warn(`RoomsService: Create Room: Room query data missing for room with name: ${newRoom.name}.`);
+            // get customException type from exceptionFactory
+            this.customException = ExceptionFactory.createException(customExceptionType.roomDataMissing);
```
Comment (slobodan-tonchevski) [inline app/src/chat/rooms-service.ts]: I dont think that you need Factories for exceptions honestly

File: app/src/chat/rooms-service.ts [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+
+        // add current user to room
+        newRoom.usersInRoom.push(currentUser._id);
+
+        try {
+            // create new Room Mongoose model and save it to DB
+            await new RoomModel({...newRoom}).save();
+            Logger.debug(`RoomsService: Create Room: New room saved to DB. RoomName = ${newRoom.name}`);
+        } catch (err) {
+            if (err instanceof Error) {
```
Comment (slobodan-tonchevski) [inline app/src/chat/rooms-service.ts]: const ALREADY_CREATED = 'E11000';
if err instaceof Error && err.message.split(' ')[0] === ALREADY_CREATED

File: app/src/chat/rooms-service.ts [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+            // get customException type from exceptionFactory
+            this.customException = ExceptionFactory.createException(customExceptionType.roomDataMissing);
+            createRoomErr = this.customException.printError();
+            return {roomName: undefined, createRoomErr};
+        }
+
+        // catch profane language in new room query
+        const badWordsFilter = new Filter();
+
+        if (badWordsFilter.isProfane(newRoom.name) || badWordsFilter.isProfane(newRoom.description)) {
```
Comment (slobodan-tonchevski) [inline app/src/chat/rooms-service.ts]: I see this is repeated into multiple methods. Please add it to a new method that will be doing these checks. Or create a completely new class where these checks will be done

File: app/src/chat/rooms-service.ts [hunk 22]
```diff
@@ -0,0 +1,378 @@
...
+    leaveRoom = async (currentUserId: Schema.Types.ObjectId| string, room: RoomPopulatedUsers): Promise<{ err: string }> => {
+        let err = '';
+        let usersInRoom: User[] | undefined;
+
+        // get currentUsersArray
+        usersInRoom = room?.usersInRoom;
+
+        // check if user is not in the current room
+        if (usersInRoom && usersInRoom.length > 0) {
+            // compare by userId, id values must be of type string because ObjectID === fails (different references)
```
Comment (slobodan-tonchevski) [inline app/src/chat/rooms-service.ts]: No need to leave that much comments. The code should be read as a book and should be underestandable.

This class is too big. I guess since some of the code is duplicated you can refactore it into smaller methods and then reuse them

File: app/src/chat/socket.ts [hunk 23]
```diff
@@ -0,0 +1,556 @@
...
+        } else {
+            // no cookie present on socketIO connection request
+            Logger.warn(`Socket: io.use: Problem authenticating user, no cookie provided socket.handshake.headers.cookie = ${socket.handshake.headers.cookie}.`);
+            const customException: CustomException = ExceptionFactory.createException(customExceptionType.unauthorizedAction);
+            next(new Error(customException.printError()));
+        }
+    }));
+
+
+    // listen to new connection and get socket instance of the connected party
```
Comment (slobodan-tonchevski) [inline app/src/chat/socket.ts]: this file could be a class and also could be refactored and spltted into chunks

File: app/src/chat/socket.ts [hunk 24]
```diff
@@ -0,0 +1,556 @@
...
+            await usersServiceSingleton.removeUsersSocketID(socket.data.userId, socket.id);
+
+            Logger.warn(`Socket: socket.on disconnect: SocketIO connection closed for socket ${socket.id}. Reason: ${reason}.`);
+        });
+    });
+}
+
+// HELPER METHODS
+// initial user checks and get specific room and currentUser data
+const initialUserRoomChecksAndDataRetrieval = async (token: string, roomName: string): Promise<{err: string, room: RoomPopulatedUsers | undefined, currentUser: User | undefined}> =>  {
```
Comment (slobodan-tonchevski) [inline app/src/chat/socket.ts]: If its a helper method, move it to separate file and export import. The file shouldnt be that big. Its doing too much logic

File: app/src/chat/users-service.ts [hunk 25]
```diff
@@ -0,0 +1,288 @@
...
+            this.customException = ExceptionFactory.createException(customExceptionType.unauthorizedAction);
+            err = this.customException.printError();
+            return {err, user: undefined};
+        }
+
+        Logger.debug(`users-service: fetchUserById: Successfully fetcher user data for user name= ${user.name}`);
+        return {err, user};
+    }
+
+    verifyUserTokenFetchUser = async (token: string): Promise<{ currentUser: User | undefined, err: string }> => {
```
Comment (slobodan-tonchevski) [inline app/src/chat/users-service.ts]: I think that you verify TOken in a wrong class. THe user service should be responsible to work with users, not validating JWT tokens

File: app/src/chat/users-service.ts [hunk 26]
```diff
@@ -0,0 +1,288 @@
...
+            this.customException = ExceptionFactory.createException(customExceptionType.unauthorizedAction);
+            verifyUserErr = this.customException.printError();
+            return {currentUser: undefined, err: verifyUserErr};
+        }
+        Logger.debug(`users-service: verifyUserTokenFetchUser: Successfully verified token, and returning user name = ${currentUser.name}`);
+        // if all is good return currentUser
+        return {currentUser, err: verifyUserErr};
+    }
+
+    removeUserFromAllRooms = async (userId: string) => {
```
Comment (slobodan-tonchevski) [inline app/src/chat/users-service.ts]: Why The user service is removing a user from a room ? Shouldnt be room service responsible for removing user from all rooms. Or maybe a room manager service ?

File: app/src/chat/users-service.ts [hunk 27]
```diff
@@ -0,0 +1,288 @@
...
+        // fetch All Rooms
+        const allRooms: Room[] = await RoomModel.find();
+
+        // check if user was in any room
+        allRoomsLoop:
+            for (const room of allRooms) {
+                // if usersInRoom array exists, check if user is in the room
```
Comment (slobodan-tonchevski) [inline app/src/chat/users-service.ts]: Too much comments and nesting. THe code is not readable. Try to chunk it into smaller parts

File: app/src/controllers/user.ts [hunk 28]
```diff
@@ -0,0 +1,140 @@
+import {Request, Response} from "express";
+import {User as UserModel} from "../db/models/user";
+import {UserNameTaken} from "../chat/exceptions/user-name-taken";
+import Logger from "../logger/logger";
+
+const signUp = async (req: Request, res: Response) => {
```
Comment (slobodan-tonchevski) [inline app/src/controllers/user.ts]: I guess that controller can be a class

File: app/src/controllers/user.ts [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+    const user = new UserModel(req.body);
```
Comment (slobodan-tonchevski) [inline app/src/controllers/user.ts]: User service should be saving. Not a controller

File: app/src/db/models/room.ts [hunk 30]
```diff
@@ -0,0 +1,64 @@
+import {model, Schema} from "mongoose";
+import {Room} from "../../interfaces/room";
+import {Message} from "../../interfaces/message";
+
+const messageSchema = new Schema<Message>({
```
Comment (slobodan-tonchevski) [inline app/src/db/models/room.ts]: each schema could go into separate file

File: app/src/db/models/user.ts [hunk 31]
```diff
@@ -0,0 +1,123 @@
...
+    ref: 'Room',
+    localField: '_id',
+    foreignField: 'author'
+});
+
+// Custom Mongoose Instance methods
+// don't use arrow function because 'this' will point to global
+userSchema.methods.generateAuthToken = async function () {
+    const user = this;
+    const token = jwt.sign({_id: user._id.toString()}, process.env.JWT_SECRET, {expiresIn: '3h'});
```
Comment (slobodan-tonchevski) [inline app/src/db/models/user.ts]: Why do we generate jwt token into model ?

Comment (slobodan-tonchevski) [reply app/src/db/models/user.ts]: There should be a JWT service generating this

File: app/src/interfaces/jwtExpPayload.ts [hunk 32]
```diff
@@ -0,0 +1,4 @@
+export interface JwtExpPayload {
```
Comment (slobodan-tonchevski) [inline app/src/interfaces/jwtExpPayload.ts]: What does jwtEXPPayload means ? 

Can it just be JwtPayload

File: app/src/logger/logger.ts [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+]
+
+const Logger = winston.createLogger({
+    level: level(),
+    levels,
+    format,
+    transports,
+})
+
+export default Logger
```
Comment (slobodan-tonchevski) [inline app/src/logger/logger.ts]: Instead of using winston directly. You could create a wrapper class logger that you can use it everywhere. THat why you dont couple your code to winston, and you can change it to other logger. Your implementation can be based on interface with strict set of rules

File: app/src/middleware/middleware.ts [hunk 34]
```diff
@@ -0,0 +1,59 @@
...
+    let exceptionMsg = '';
+    let token: string;
+    if (req.cookies && req.cookies.access_token) {
+        // retrieve cookie access_token
+        token = req.cookies.access_token;
+        try {
+            // cast decodedToken to UserTokenPayload
+            const decodedToken = (jwt.verify(token, process.env.JWT_SECRET)) as UserTokenPayload;
+            // find user by using the _id from the token
+            const user = await User.findOne({_id: decodedToken._id, 'tokens.token': token});
```
Comment (slobodan-tonchevski) [inline app/src/middleware/middleware.ts]: we can do these checks into a service. WE dont want to access models from middleware directly. But use a model that will be doing this checks and just returning true/false or throwing exception if something is not valid

File: .gitignore (hunk 1 above)
Comment (kikooOOoo16) [reply .gitignore]: Agreed, it is not a guarantee that a dev that will look at this code uses the same IDE and most of the IDE stuff will be auto generated anyway. Source code should probably hold only necessary files for building and running the app. Removed .idea directory.

File: app/server.ts (hunk 4 above)
Comment (kikooOOoo16) [reply app/server.ts]: App crashes with TypeError [ERR_INVALID_OPT_VALUE('options', option)], should I gracefully close the app if normalizing port fails, with an err log ?

File: app/custom.d.ts (hunk 3 above)
Comment (kikooOOoo16) [reply app/custom.d.ts:1]: Needed to extend the type information for the Express module in order to include the token value (token value not needed anymore) and the user value which is passed from the express middleware (if successfully validated). Was not sure what to name it and if I could place it somewhere better. What is the convention for "custom" d.ts files ?

File: app/src/db/models/room.ts (hunk 30 above)
Comment (kikooOOoo16) [reply app/src/db/models/room.ts:5]: Agreed, separated schemas into separate files.

File: app/src/interfaces/jwtExpPayload.ts (hunk 32 above)
Comment (kikooOOoo16) [reply app/src/interfaces/jwtExpPayload.ts]: Leftover interface from previous version, naming is bad indeed, removed the interface as it was no longer needed.

File: app/src/chat/rooms-service.ts (hunk 20 above)
Comment (kikooOOoo16) [reply app/src/chat/rooms-service.ts:66]: Should the const be declared inside the err function or should it be set as a readonly class variable then called inside err function ?

File: app/server.ts (hunk 5 above)
Comment (kikooOOoo16) [reply app/server.ts]: Should be const addr = server.adress(), I've forgotten to add it...

File: app/src/chat/exceptions/custom-exception.ts (hunk 9 above)
Comment (kikooOOoo16) [reply app/src/chat/exceptions/custom-exception.ts]: TypeScript 2.1+ seems to have had breaking changes regarding extending built in Classes/Interfaces like Error. To be honest I didn't know that, was following a guide on how to create custom exceptions extending Error. https://github.com/Microsoft/TypeScript-wiki/blob/main/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work

File: app/src/chat/exceptions/expired-user-token.ts (hunk 12 above)
Comment (kikooOOoo16) [reply app/src/chat/exceptions/expired-user-token.ts]: Not using custom errors properly, should be throwing and catching them... will refactor.

File: app/src/chat/rooms-service.ts (hunk 18 above)
Comment (kikooOOoo16) [reply app/src/chat/rooms-service.ts]: Would you mind explaining this in a little more detail ? Seems useful. Never mind learned about DTO classes in NestJS.

File: app/src/db/models/user.ts (hunk 31 above)
Comment (kikooOOoo16) [reply app/src/db/models/user.ts]: Created an Auth service that takes care of anything related to authentication.

File: app/src/logger/logger.ts (hunk 33 above)
Comment (kikooOOoo16) [reply app/src/logger/logger.ts]: Some resources about this would be useful, would love to implement it.

File: app/src/chat/socket.ts (hunk 23 above)
Comment (kikooOOoo16) [reply app/src/chat/socket.ts]: Removed helper methods and other logic that shouldn't be in here but was unable to restructure the io.on(connection) method as a whole, will change it once I find a good way to do it...
